File: results/0.5/HTTP-1.1/http-1.1-meta-llama-llama-3.3-70b-instruct-turbo.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define HAMMER_NO_PARSER

typedef struct {
    char* method;
    char* request_uri;
    char* http_version;
} request_line_t;

typedef struct {
    char* accept;
    char* accept_charset;
    char* accept_encoding;
    char* accept_language;
    char* authorization;
    char* cache_control;
    char* connection;
    char* content_length;
    char* content_type;
    char* date;
    char* expect;
    char* from;
    char* host;
    char* if_match;
    char* if_modified_since;
    char* if_none_match;
    char* if_range;
    char* if_unmodified_since;
    char* max_forwards;
    char* proxy_authorization;
    char* range;
    char* referer;
    char* te;
    char* user_agent;
} request_headers_t;

typedef struct {
    char* http_version;
    int status_code;
    char* reason_phrase;
} status_line_t;

typedef struct {
    char* accept_ranges;
    char* age;
    char* allow;
    char* cache_control;
    char* connection;
    char* content_encoding;
    char* content_language;
    char* content_length;
    char* content_location;
    char* content_md5;
    char* content_range;
    char* content_type;
    char* date;
    char* etag;
    char* expires;
    char* last_modified;
    char* location;
    char* proxy_authenticate;
    char* retry_after;
    char* server;
    char* set_cookie;
    char* trailer;
    char* transfer_encoding;
    char* upgrade;
    char* vary;
    char* via;
    char* warning;
    char* www_authenticate;
} response_headers_t;

typedef struct {
    request_line_t request_line;
    request_headers_t request_headers;
    status_line_t status_line;
    response_headers_t response_headers;
} http_message_t;

HParser* request_line_parser;
HParser* request_headers_parser;
HParser* status_line_parser;
HParser* response_headers_parser;
HParser* http_message_parser;

void init_parsers() {
    request_line_parser = h_sequence(
        h_string("GET"),
        h_string(" "),
        h_regex("[^ ]+"),
        h_string(" "),
        h_regex("HTTP/1\\.1"),
        h_eol,
        &(request_line_t){
            .method = h_capture(1),
            .request_uri = h_capture(2),
            .http_version = h_capture(3)
        }
    );

    request_headers_parser = h_sequence(
        h_optional(
            h_sequence(
                h_string("Accept: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .accept = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Accept-Charset: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .accept_charset = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Accept-Encoding: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .accept_encoding = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Accept-Language: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .accept_language = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Authorization: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .authorization = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Cache-Control: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .cache_control = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Connection: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .connection = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Content-Length: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .content_length = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Content-Type: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .content_type = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Date: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .date = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Expect: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .expect = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("From: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .from = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Host: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .host = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("If-Match: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .if_match = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("If-Modified-Since: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .if_modified_since = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("If-None-Match: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .if_none_match = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("If-Range: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .if_range = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("If-Unmodified-Since: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .if_unmodified_since = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Max-Forwards: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .max_forwards = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Proxy-Authorization: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .proxy_authorization = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Range: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .range = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Referer: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .referer = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("TE: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .te = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("User-Agent: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(request_headers_t){
                    .user_agent = h_capture(1)
                }
            )
        ),
        h_eol
    );

    status_line_parser = h_sequence(
        h_string("HTTP/1.1 "),
        h_regex("\\d{3}"),
        h_string(" "),
        h_regex("[^\\r\\n]+"),
        h_eol,
        &(status_line_t){
            .http_version = h_capture(0),
            .status_code = atoi(h_capture(1)),
            .reason_phrase = h_capture(2)
        }
    );

    response_headers_parser = h_sequence(
        h_optional(
            h_sequence(
                h_string("Accept-Ranges: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .accept_ranges = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Age: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .age = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Allow: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .allow = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Cache-Control: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .cache_control = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Connection: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .connection = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Content-Encoding: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .content_encoding = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Content-Language: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .content_language = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Content-Length: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .content_length = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Content-Location: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .content_location = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Content-MD5: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .content_md5 = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Content-Range: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .content_range = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Content-Type: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .content_type = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Date: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .date = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("ETag: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .etag = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Expires: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .expires = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Last-Modified: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .last_modified = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Location: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .location = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Proxy-Authenticate: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .proxy_authenticate = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Retry-After: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .retry_after = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Server: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .server = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Set-Cookie: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .set_cookie = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Trailer: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .trailer = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Transfer-Encoding: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .transfer_encoding = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Upgrade: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .upgrade = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Vary: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .vary = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Via: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .via = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("Warning: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .warning = h_capture(1)
                }
            )
        ),
        h_optional(
            h_sequence(
                h_string("WWW-Authenticate: "),
                h_regex("[^\\r\\n]+"),
                h_eol,
                &(response_headers_t){
                    .www_authenticate = h_capture(1)
                }
            )
        ),
        h_eol
    );

    http_message_parser = h_sequence(
        request_line_parser,
        request_headers_parser,
        status_line_parser,
        response_headers_parser,
        &(http_message_t){
            .request_line = h_capture(0),
            .request_headers = h_capture(1),
            .status_line = h_capture(2),
            .response_headers = h_capture(3)
        }
    );
}

int main(int argc, char** argv) {
    init_parsers();

    if (argc != 2) {
        printf("Usage: %s <file>\n", argv[0]);
        return 1;
    }

    FILE* file = fopen(argv[1], "rb");
    if (!file) {
        printf("Failed to open file %s\n", argv[1]);
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);

    uint8_t* buffer = malloc(file_size);
    if (!buffer) {
        printf("Failed to allocate memory\n");
        fclose(file);
        return 1;
    }

    size_t bytes_read = fread(buffer, 1, file_size, file);
    if (bytes_read != file_size) {
        printf("Failed to read file %s\n", argv[1]);
        free(buffer);
        fclose(file);
        return 1;
    }

    fclose(file);

    http_message_t message;
    HParseResult* result = h_parse(http_message_parser, buffer, file_size);
    if (result->status != H_OK) {
        printf("Failed to parse HTTP message\n");
        free(buffer);
        return 1;
    }

    printf("Request Line:\n");
    printf("  Method: %s\n", message.request_line.method);
    printf("  Request URI: %s\n", message.request_line.request_uri);
    printf("  HTTP Version: %s\n", message.request_line.http_version);

    printf("Request Headers:\n");
    if (message.request_headers.accept) printf("  Accept: %s\n", message.request_headers.accept);
    if (message.request_headers.accept_charset) printf("  Accept-Charset: %s\n", message.request_headers.accept_charset);
    if (message.request_headers.accept_encoding) printf("  Accept-Encoding: %s\n", message.request_headers.accept_encoding);
    if (message.request_headers.accept_language) printf("  Accept-Language: %s\n", message.request_headers.accept_language);
    if (message.request_headers.authorization) printf("  Authorization: %s\n", message.request_headers.authorization);
    if (message.request_headers.cache_control) printf("  Cache-Control: %s\n", message.request_headers.cache_control);
    if (message.request_headers.connection) printf("  Connection: %s\n", message.request_headers.connection);
    if (message.request_headers.content_length) printf("  Content-Length: %s\n", message.request_headers.content_length);
    if (message.request_headers.content_type) printf("  Content

File: results/0.5/HTTP-1.1/http-1.1-meta-llama-llama-3.3-70b-instruct-turbo.rs
use nom::{
    branch::alt,
    bytes::complete::{tag, take_while1},
    character::complete::{char, digit1, multispace0, multispace1},
    combinator::{map, map_res, value},
    multi::{many0, separated_list0},
    sequence::{delimited, preceded, tuple},
    IResult,
};
use std::env;
use std::fs::File;
use std::io::Read;
use std::str;

#[derive(Debug, PartialEq, Clone)]
enum HttpMethod {
    GET,
    HEAD,
    POST,
    PUT,
    DELETE,
    CONNECT,
    OPTIONS,
    TRACE,
}

fn http_method(input: &[u8]) -> IResult<&[u8], HttpMethod> {
    alt((
        value(HttpMethod::GET, tag("GET")),
        value(HttpMethod::HEAD, tag("HEAD")),
        value(HttpMethod::POST, tag("POST")),
        value(HttpMethod::PUT, tag("PUT")),
        value(HttpMethod::DELETE, tag("DELETE")),
        value(HttpMethod::CONNECT, tag("CONNECT")),
        value(HttpMethod::OPTIONS, tag("OPTIONS")),
        value(HttpMethod::TRACE, tag("TRACE")),
    ))(input)
}

#[derive(Debug, PartialEq)]
enum HttpVersion {
    HTTP11,
}

fn http_version(input: &[u8]) -> IResult<&[u8], HttpVersion> {
    map_res(
        tag("HTTP/1.1"),
        |_| Ok(HttpVersion::HTTP11),
    )(input)
}

#[derive(Debug, PartialEq)]
enum HttpStatus {
    Informational(u16),
    Successful(u16),
    Redirection(u16),
    ClientError(u16),
    ServerError(u16),
}

fn http_status(input: &[u8]) -> IResult<&[u8], HttpStatus> {
    let (input, code) = map_res(digit1, |s: &[u8]| str::from_utf8(s).unwrap().parse::<u16>())(input)?;
    match code {
        100..=199 => Ok((input, HttpStatus::Informational(code))),
        200..=299 => Ok((input, HttpStatus::Successful(code))),
        300..=399 => Ok((input, HttpStatus::Redirection(code))),
        400..=499 => Ok((input, HttpStatus::ClientError(code))),
        500..=599 => Ok((input, HttpStatus::ServerError(code))),
        _ => Err(nom::Err::Error(nom::error::ErrorKind::Digit)),
    }
}

#[derive(Debug, PartialEq)]
enum HttpHeader {
    Accept(String),
    AcceptCharset(String),
    AcceptEncoding(String),
    AcceptLanguage(String),
    Authorization(String),
    CacheControl(String),
    Connection(String),
    ContentLength(u64),
    ContentType(String),
    Date(String),
    Expect(String),
    From(String),
    Host(String),
    IfMatch(String),
    IfModifiedSince(String),
    IfNoneMatch(String),
    IfRange(String),
    IfUnmodifiedSince(String),
    MaxForwards(u16),
    ProxyAuthorization(String),
    Range(String),
    Referer(String),
    TE(String),
    UserAgent(String),
    Via(String),
    Warning(String),
    WWWAuthenticate(String),
}

fn http_header(input: &[u8]) -> IResult<&[u8], HttpHeader> {
    alt((
        map(preceded(tag("Accept: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::Accept(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("Accept-Charset: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::AcceptCharset(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("Accept-Encoding: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::AcceptEncoding(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("Accept-Language: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::AcceptLanguage(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("Authorization: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::Authorization(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("Cache-Control: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::CacheControl(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("Connection: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::Connection(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("Content-Length: "), map_res(digit1, |s: &[u8]| str::from_utf8(s).unwrap().parse::<u64>())), |n| HttpHeader::ContentLength(n)),
        map(preceded(tag("Content-Type: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::ContentType(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("Date: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::Date(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("Expect: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::Expect(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("From: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::From(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("Host: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::Host(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("If-Match: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::IfMatch(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("If-Modified-Since: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::IfModifiedSince(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("If-None-Match: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::IfNoneMatch(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("If-Range: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::IfRange(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("If-Unmodified-Since: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::IfUnmodifiedSince(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("Max-Forwards: "), map_res(digit1, |s: &[u8]| str::from_utf8(s).unwrap().parse::<u16>())), |n| HttpHeader::MaxForwards(n)),
        map(preceded(tag("Proxy-Authorization: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::ProxyAuthorization(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("Range: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::Range(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("Referer: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::Referer(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("TE: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::TE(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("User-Agent: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::UserAgent(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("Via: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::Via(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("Warning: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::Warning(String::from_utf8_lossy(s).into_owned())),
        map(preceded(tag("WWW-Authenticate: "), take_while1(|c| c != b'\r' && c != b'\n')), |s| HttpHeader::WWWAuthenticate(String::from_utf8_lossy(s).into_owned())),
    ))(input)
}

#[derive(Debug, PartialEq)]
enum HttpRequest {
    RequestLine(HttpMethod, String, HttpVersion),
    Headers(Vec<HttpHeader>),
}

fn http_request(input: &[u8]) -> IResult<&[u8], HttpRequest> {
    alt((
        map(
            tuple((
                http_method,
                multispace1,
                take_while1(|c| c != b' ' && c != b'\r' && c != b'\n'),
                multispace1,
                http_version,
                tag("\r\n"),
            )),
            |(method, _, path, _, version, _)| HttpRequest::RequestLine(method, String::from_utf8_lossy(path).into_owned(), version),
        ),
        map(
            many0(alt((http_header, map(tag("\r\n"), |_| ())))),
            |headers| HttpRequest::Headers(headers.into_iter().filter_map(|h| h).collect()),
        ),
    ))(input)
}

#[derive(Debug, PartialEq)]
enum HttpResponse {
    StatusLine(HttpStatus, String),
    Headers(Vec<HttpHeader>),
}

fn http_response(input: &[u8]) -> IResult<&[u8], HttpResponse> {
    alt((
        map(
            tuple((
                http_version,
                multispace1,
                http_status,
                multispace1,
                take_while1(|c| c != b'\r' && c != b'\n'),
                tag("\r\n"),
            )),
            |(version, _, status, _, reason, _)| HttpResponse::StatusLine(status, String::from_utf8_lossy(reason).into_owned()),
        ),
        map(
            many0(alt((http_header, map(tag("\r\n"), |_| ())))),
            |headers| HttpResponse::Headers(headers.into_iter().filter_map(|h| h).collect()),
        ),
    ))(input)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let file_path = &args[1];
    let mut file = File::open(file_path).unwrap();
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer).unwrap();
    match http_request(&buffer) {
        Ok((_remaining, request)) => println!("{:?}", request),
        Err(e) => println!("{:?}", e),
    }
    match http_response(&buffer) {
        Ok((_remaining, response)) => println!("{:?}", response),
        Err(e) => println!("{:?}", e),
    }
}